### PR TITLE
add dpr evaluation files

### PR DESCRIPTION
**Pull request by MXueguang**
for now I have two files in `scripts/dpr`

1. `retrieval.py`:
```
python retrieval.py --dense-index <dpr wiki index> \
                    --spares-index <sindex that has raw> \
                    --encoder <dpr_encoder> \
                    --qas <question.answers.csv> \
                    --output <json file>
```
We will get a json file that has the following structure by this:
```
{
    0: {
            "question": <question text>
            "answers": [<ans1>, <ans2>]
            "contexts":[
                                 {"docid": <id>, "score": <score>, "text": <text>},
                       ]
        }
}
```

2. `evaluate.py`: will read the json file and calculate accuracy.

### CODE REVIEW

**Comment by lintool** [comment]
This is e2e F1/EM/etc, right? I think this should be in pygaggle? 
The scripts in pyserini should only be evaluating retriever effectiveness?

**Comment by MXueguang** [comment]
The retriever effectiveness is evaluated by if the answer is in retrieved docs.
i.e. _measured as the percentage of top 20/100 retrieved passages that contain the answer_
so we need the EM here.

**Comment by lintool** [comment]
I see. Why can't use just use https://github.com/castorini/pyserini/pull/330 and add `--dpr` as another output format?

**Comment by MXueguang** [comment]
Dpr is not hybrid rn, maybe add to here first? https://github.com/castorini/pyserini/blob/master/pyserini/dsearch/__main__.py
use the `--dpr` to control the output format.
but we might need a `--corpus` option to take the sparse index that contains the raw text for dpr.
we also need to add the dpr question-answers to topics in this case?


maybe another solution:
retrieve dpr the same as msmarco etc. i.e. we get a msmarco or trec style run file.
then we use a script to convert the run file to dpr retrieval file. (i.e. pull raw text from corpus)

**Comment by lintool** [comment]
> Dpr is not hybrid rn, maybe add to here first? https://github.com/castorini/pyserini/blob/master/pyserini/dsearch/__main__.py

+1

> maybe another solution:
> retrieve dpr the same as msmarco etc. i.e. we get a msmarco or trec style run file.
> then we use a script to convert the run file to dpr retrieval file. (i.e. pull raw text from corpus)

yes, I like this solution - use our existing pipeline, but specialized DPR-specific "post-processing".

**Comment by MXueguang** [comment]
removed the redundant `--encoded-queries` in this PR as well

File: docs/dense-retrieval.md [hunk 1]
```diff
@@ -167,6 +162,45 @@ map                     all     0.3706
...
+$ python -m pyserini.dsearch --topics nq-dev.qa.json \
+                             --index wikipedia-dpr-hnsw \
+                             --encoder facebook/dpr-question_encoder-single-nq-base \
+                             --output runs/run.dpr.hnsw.trec 
+```
+
+Convert the TREC style run file to retrieval result file in `json` format
+```bash
+$ python -m scripts.dpr.convert_trec_run_to_retrieval_json --qas nq-dev.qa.json \
+                                                         --index lucene_index_wikipedia_psgs_w100_with_options \
```
Comment (lintool) [inline docs/dense-retrieval.md:195]: alignment?

File: docs/dense-retrieval.md [hunk 2]
```diff
@@ -167,6 +162,45 @@ map                     all     0.3706
...
+
+
+Download question&answers from Natural Question dev set
+```bash
+$ wget https://dl.fbaipublicfiles.com/dpr/data/retriever/nq-dev.qa.csv
+```
+
+Convert the `nq-dev.qa.csv` to topics file in `json` format
+```bash
+$ python scripts/dpr/convert_qas_csv_to_topic_json.json --input nq-dev.qa.csv --output nq-dev.qa.json
```
Comment (lintool) [inline docs/dense-retrieval.md:181]: Just a note for future reference, maybe this should be added on the anserini end? Okay like this for now.

File: docs/dense-retrieval.md [hunk 3]
```diff
@@ -167,6 +162,45 @@ map                     all     0.3706
...
+                                                         --output retrieval.json
+```
+
+Evaluate
+```bash
+$ python scripts/dpr/evaluate.py --retrieval retrieval.json --topk 20
+Top20  accuracy: 0.7777777777777778
+$ python scripts/dpr/evaluate.py --retrieval retrieval.json --topk 100
+Top100 accuracy: 0.8442388945986068
+```
```
Comment (lintool) [inline docs/dense-retrieval.md:206]: Add in the figure in the book for comparison.

**Review by lintool — APPROVED** [review]
minor comments, otherwise lgtm

File: docs/dense-retrieval.md (hunk 2 above)
Comment (MXueguang) [reply docs/dense-retrieval.md:181]: yeah

Comment (lintool) [reply docs/dense-retrieval.md:181]: Maybe this is a task for @KaiSun314 ?

**Pull request merged**